File: apps/api/app/models/accelerator.py
"""Pydantic models for application accelerator operations."""
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


# Request Models
class ResumeOptimizeRequest(BaseModel):
    """Request to optimize resume."""

    resume_id: str = Field(..., min_length=1)
    job_id: str = Field(..., min_length=1)


class CoverLetterGenerateRequest(BaseModel):
    """Request to generate cover letter."""

    resume_id: str = Field(..., min_length=1)
    job_id: str = Field(..., min_length=1)
    tone: str = Field(default="professional")
    length: str = Field(default="medium")


class RecruiterOutreachRequest(BaseModel):
    """Request to generate recruiter outreach."""

    resume_id: str = Field(..., min_length=1)
    job_id: Optional[str] = None
    message_type: str = Field(default="linkedin_connection")
    tone: str = Field(default="professional")
    length: str = Field(default="medium")


class InterviewGenerateRequest(BaseModel):
    """Request to generate interview kit."""

    resume_id: str = Field(..., min_length=1)
    job_id: str = Field(..., min_length=1)


class DocumentExportRequest(BaseModel):
    """Request to export document."""

    document_id: str = Field(..., min_length=1)
    format: str = Field(default="markdown")


# Response Models
class ResumeChange(BaseModel):
    """Resume change detail."""

    original: str
    optimized: str
    reason: str
    ats_improvement: str
    recruiter_impact: str


class SkillsOptimization(BaseModel):
    """Skills optimization detail."""

    original_order: list[str]
    optimized_order: list[str]
    reasoning: str


class KeywordAddition(BaseModel):
    """Keyword addition detail."""

    keyword: str
    where_added: str
    reason: str


class ResumeOptimizeResponse(BaseModel):
    """Resume optimization response."""

    id: str
    optimized_summary: str
    optimized_experience: list[ResumeChange]
    optimized_skills: SkillsOptimization
    optimized_projects: list[ResumeChange]
    keyword_additions: list[KeywordAddition]
    optimization_score: int
    estimated_ats_improvement: int
    estimated_match_increase: int
    estimated_interview_probability: int
    created_at: datetime


class PersonalizationPoint(BaseModel):
    """Personalization point."""

    point: str
    source: str


class CoverLetterResponse(BaseModel):
    """Cover letter response."""

    id: str
    content: str
    tone: str
    length: str
    company_name: str
    role_title: str
    personalization_points: list[PersonalizationPoint]
    created_at: datetime


class RecruiterMessageResponse(BaseModel):
    """Recruiter message response."""

    id: str
    message_type: str
    subject: Optional[str]
    content: str
    tone: str
    length: str
    personalization_reason: str
    call_to_action: str
    recipient_name: Optional[str]
    recipient_company: Optional[str]
    platform: str
    created_at: datetime


class TechnicalTopic(BaseModel):
    """Technical topic."""

    topic: str
    priority: str
    resources: list[str]


class BehavioralQuestion(BaseModel):
    """Behavioral question."""

    question: str
    star_suggestion: str
    priority: str


class ProjectQuestion(BaseModel):
    """Project question."""

    question: str
    suggested_answer: str
    priority: str


class CodingTopic(BaseModel):
    """Coding topic."""

    topic: str
    priority: str
    practice_problems: list[str]


class SystemDesignTopic(BaseModel):
    """System design topic."""

    topic: str
    priority: str
    key_concepts: list[str]


class QuestionToAsk(BaseModel):
    """Question to ask interviewer."""

    question: str
    reason: str


class StudyPlanItem(BaseModel):
    """Study plan item."""

    time: Optional[str] = None
    task: str
    priority: str


class StudyPlanDay(BaseModel):
    """Study plan day."""

    day: str
    tasks: list[str]
    focus: str


class PriorityRanking(BaseModel):
    """Priority ranking."""

    highest_priority: list[str]
    high_priority: list[str]
    medium_priority: list[str]
    low_priority: list[str]


class InterviewKitResponse(BaseModel):
    """Interview kit response."""

    id: str
    company_name: str
    role_title: str
    company_overview: str
    role_overview: str
    responsibilities: list[str]
    technical_topics: list[TechnicalTopic]
    behavioral_questions: list[BehavioralQuestion]
    project_questions: list[ProjectQuestion]
    resume_questions: list[ProjectQuestion]
    coding_topics: list[CodingTopic]
    system_design_topics: list[SystemDesignTopic]
    hr_questions: list[dict[str, str]]
    salary_tips: list[str]
    questions_to_ask: list[QuestionToAsk]
    study_plan_90min: list[StudyPlanItem]
    study_plan_3day: list[StudyPlanDay]
    study_plan_7day: list[StudyPlanDay]
    priority_ranking: PriorityRanking
    created_at: datetime


class GeneratedDocumentResponse(BaseModel):
    """Generated document response."""

    id: str
    document_type: str
    title: str
    description: Optional[str]
    format: str
    file_path: Optional[str]
    file_size: Optional[int]
    export_count: int
    created_at: datetime


class DocumentExportResponse(BaseModel):
    """Document export response."""

    file_path: str
    format: str
    file_size: int

File: apps/web/src/hooks/api/accelerator.ts
import { useMutation, useQueryClient } from '@tanstack/react-query';
import apiClient from '@/lib/api-client';

// TypeScript interfaces
export interface ResumeChange {
  original: string;
  optimized: string;
  reason: string;
  ats_improvement: string;
  recruiter_impact: string;
}

export interface SkillsOptimization {
  original_order: string[];
  optimized_order: string[];
  reasoning: string;
}

export interface KeywordAddition {
  keyword: string;
  where_added: string;
  reason: string;
}

export interface ResumeOptimization {
  id: string;
  optimized_summary: string;
  optimized_experience: ResumeChange[];
  optimized_skills: SkillsOptimization;
  optimized_projects: ResumeChange[];
  keyword_additions: KeywordAddition[];
  optimization_score: number;
  estimated_ats_improvement: number;
  estimated_match_increase: number;
  estimated_interview_probability: number;
  created_at: string;
}

export interface PersonalizationPoint {
  point: string;
  source: string;
}

export interface CoverLetter {
  id: string;
  content: string;
  tone: string;
  length: string;
  company_name: string;
  role_title: string;
  personalization_points: PersonalizationPoint[];
  created_at: string;
}

export interface RecruiterMessage {
  id: string;
  message_type: string;
  subject: string | null;
  content: string;
  tone: string;
  length: string;
  personalization_reason: string;
  call_to_action: string;
  recipient_name: string | null;
  recipient_company: string | null;
  platform: string;
  created_at: string;
}

export interface TechnicalTopic {
  topic: string;
  priority: string;
  resources: string[];
}

export interface BehavioralQuestion {
  question: string;
  star_suggestion: string;
  priority: string;
}

export interface ProjectQuestion {
  question: string;
  suggested_answer: string;
  priority: string;
}

export interface CodingTopic {
  topic: string;
  priority: string;
  practice_problems: string[];
}

export interface SystemDesignTopic {
  topic: string;
  priority: string;
  key_concepts: string[];
}

export interface QuestionToAsk {
  question: string;
  reason: string;
}

export interface StudyPlanItem {
  time: string | null;
  task: string;
  priority: string;
}

export interface StudyPlanDay {
  day: string;
  tasks: string[];
  focus: string;
}

export interface PriorityRanking {
  highest_priority: string[];
  high_priority: string[];
  medium_priority: string[];
  low_priority: string[];
}

export interface InterviewKit {
  id: string;
  company_name: string;
  role_title: string;
  company_overview: string;
  role_overview: string;
  responsibilities: string[];
  technical_topics: TechnicalTopic[];
  behavioral_questions: BehavioralQuestion[];
  project_questions: ProjectQuestion[];
  resume_questions: ProjectQuestion[];
  coding_topics: CodingTopic[];
  system_design_topics: SystemDesignTopic[];
  hr_questions: { question: string; suggested_answer: string }[];
  salary_tips: string[];
  questions_to_ask: QuestionToAsk[];
  study_plan_90min: StudyPlanItem[];
  study_plan_3day: StudyPlanDay[];
  study_plan_7day: StudyPlanDay[];
  priority_ranking: PriorityRanking;
  created_at: string;
}

// API hooks
export function useOptimizeResume() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (data: { resume_id: string; job_id: string }) =>
      apiClient.post<ResumeOptimization>('/api/v1/accelerator/optimize/resume', data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['accelerator'] });
    },
  });
}

export function useGenerateCoverLetter() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (data: { resume_id: string; job_id: string; tone?: string; length?: string }) =>
      apiClient.post<CoverLetter>('/api/v1/accelerator/cover-letter/generate', data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['accelerator'] });
    },
  });
}

export function useGenerateOutreach() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (data: {
      resume_id: string;
      job_id?: string;
      message_type?: string;
      tone?: string;
      length?: string;
    }) => apiClient.post<RecruiterMessage>('/api/v1/accelerator/outreach/generate', data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['accelerator'] });
    },
  });
}

export function useGenerateInterviewKit() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (data: { resume_id: string; job_id: string }) =>
      apiClient.post<InterviewKit>('/api/v1/accelerator/interview/generate', data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['accelerator'] });
    },
  });
}

export function useExportDocument() {
  return useMutation({
    mutationFn: (data: { document_id: string; format?: string }) =>
      apiClient.post<{ file_path: string; format: string; file_size: number }>(
        '/api/v1/accelerator/documents/export',
        data
      ),
  });
}
